ipatests/test_integration/test_epn.py

```diff
@@ -59,6 +63,85 @@ def postconf(host, option):
     host.run_command(r"postconf -e '%s'" % option)
 
 
+def generate_user(host, uid, **kwargs):
+    user_defaults = {
+        'objectClass': [
+            'ipaobject',
+            'person',
+            'top',
+            'ipasshuser',
+            'inetorgperson',
+            'organizationalperson',
+            'krbticketpolicyaux',
+            'krbprincipalaux',
+            'inetuser',
+            'posixaccount',
+            'ipaSshGroupOfPubKeys',
+            'mepOriginEntry'],
+        'ipaUniqueID': ['autogenerate'],
+        'loginShell': ['/bin/zsh'],
+        'uidNumber': ['-1'],
+        'gidNumber': ['-1'],
+    }
+
+    user = dict(user_defaults)
+    user['dn'] = 'uid={uid},cn=users,cn=accounts,{suffix}'.format(
+        uid=uid,
+        suffix=host.hostname,
+    )
+    user['uid'] = [uid]
+    user['displayName'] = ['{0} {0}'.format(uid)]
+    user['initials'] = ['{}{}'.format(uid[1], uid[-1])]
+    user['gecos'] = user['displayName']
+    user['sn'] = [uid]
+    user['homeDirectory'] = ['/other-home/{}'.format(uid)]
+    user['mail'] = ['{uid}@{domain}'.format(
+        uid=uid, domain=host.domain)]
+    user['krbPrincipalName'] = ['{uid}@{realm}'.format(
+        uid=uid, realm=host.domain.realm)]
+    user['krbCanonicalName'] = ['{uid}@{realm}'.format(
+        uid=uid, realm=host.domain.realm)]
+    user['givenName'] = [uid]
+    user['cn'] = ['{0} {0}'.format(uid)]
+
+    if "krbPasswordExpiration" in kwargs.keys():
+        user['krbPasswordExpiration'] = kwargs['krbPasswordExpiration']
+
+    if "userPassword" in kwargs.keys():
+        user['userPassword'] = kwargs['userPassword']
+
+    if "Mail2" in kwargs.keys():
+        user['mail'].append(kwargs['Mail2'])
+
+    if "Mail3" in kwargs.keys():
+        user['mail'].append(kwargs['Mail3'])
+
+    return user
+
+
+def generate_user_ldif(user_dict):
+    writer = ldif.LDIFWriter(io.StringIO(), None, 76, '\n')
+    user_dict_bytes = {
+        key: val.encode('utf-8') if not isinstance(val, list)
+        else [item.encode('utf-8') for item in val]
+        for key, val in user_dict.items()
+    }
+    return writer.unparse(dn=user_dict['dn'], record=user_dict_bytes)
```

Comment:
unparse is not a function that returns data. It should be used like this:
```
io = StringIO()
writer = ldif.LDIFWriter(io)
user_dict_bytes = {
    key: val.encode('utf-8') if not isinstance(val, list)
    else [item.encode('utf-8') for item in val]
    for key, val in user_dict.items()
}
writer.unparse(dn=user_dict['dn'], record=user_dict_bytes)
return io.getvalue()
```

Reply:
Thank you Flo!